File: 25.reverse-nodes-in-k-group.go
/*
 * @lc app=leetcode id=25 lang=golang
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || k <= 1 {
		return head
	}
	preHead := new(ListNode)
	preHead.Next = head
	prev := preHead
	for true {
		last := prev
		n := 0
		for n < k {
			n += 1
			last = last.Next
			if last == nil {
				return preHead.Next
			}
		}
		n = 0
		last = prev.Next
		for n < k-1 {
			n += 1
			first := prev.Next
			prev.Next = last.Next
			last.Next = last.Next.Next
			prev.Next.Next = first

			n := preHead.Next
			for n != nil {
				n = n.Next
			}
		}
		prev = last
	}

	return preHead.Next
}

// @lc code=end

File: 25.reverse-nodes-in-k-group.java
/*
 * @lc app=leetcode id=25 lang=java
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (k == 1) return head;
        if (head == null) return head;
        ListNode hair = new ListNode();
        ListNode pre = hair, foot = head;
        while (head != null) {
            int n = k;
            for (; n > 1 && foot.next != null; n--) {
                foot = foot.next;
            }
            if (n > 1) break;
            ListNode next = foot.next;
            reverse(head, foot);
            pre.next = foot;
            pre = head;
            head = next;
            foot = next;
        }
        pre.next = head;

        return hair.next;
    }

    private void reverse(ListNode first, ListNode last) {
        ListNode next = first.next;
        while (first != last) {
            ListNode nnext = next.next;
            next.next = first;
            first = next;
            next = nnext;
        }
    }
}
// @lc code=end

